File: RepoTransBench/source_projects/Python/powergo_pytracking/public_tests/test_pytracking_public.py
from pytracking import (
    Configuration, get_click_tracking_url, get_click_tracking_url_path,
    get_open_tracking_url_path,
    get_click_tracking_result, get_open_tracking_pixel, get_open_tracking_url,
    get_open_tracking_result)

ALT_URL_TO_TRACK = "https://anotherdomain.io/tracking/?data=newdata"

ALT_BASE_CLICK_TRACKING_URL = "https://y.z.com/track/"

ALT_BASE_OPEN_TRACKING_URL = "https://y.z.com/track/open/"

ALT_ENCRYPTION_KEY = b'KJhVbRLnoCIPBLB7UfPIW3vOTsmUNxCZ-2inOtTdu6UV='

ALT_METADATA = {
    "param4": "val4",
    "another": True,
    "nested_alt": {"paramA": "valA"}}

ALT_DEFAULT_METADATA = {
    "key42": False,
    "strangeé": "winoèèè",
    "paramX": "other3"
}

ALT_EXPECTED_METADATA = {}
ALT_EXPECTED_METADATA.update(ALT_DEFAULT_METADATA)
ALT_EXPECTED_METADATA.update(ALT_METADATA)

ALT_WEBHOOK_URL = "https://notify.me/webhook/"

ALT_REQUEST_DATA = {
    "user_agent": "Safari",
    "user_ip": "192.168.1.1"
}

ALT_CONFIGURATION = Configuration(
    webhook_url=ALT_WEBHOOK_URL,
    base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL,
    base_click_tracking_url=ALT_BASE_CLICK_TRACKING_URL,
    default_metadata=ALT_DEFAULT_METADATA)

def test_public_get_open_tracking_pixel():
    (pixel, mime) = get_open_tracking_pixel()
    assert isinstance(pixel, bytes)
    assert mime == "image/png"

def test_public_basic_get_open_tracking_url():
    url = get_open_tracking_url(
        base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL)
    assert url.startswith(ALT_BASE_OPEN_TRACKING_URL)
    assert url[len(ALT_BASE_OPEN_TRACKING_URL):] != ""  # ensure there's something encoded

def test_public_basic_get_open_tracking_url_append_slash():
    url = get_open_tracking_url(
        base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL,
        append_slash=True)
    assert url.endswith("/")

def test_public_in_config_open_tracking_url():
    url = get_open_tracking_url(
        base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL,
        metadata=ALT_METADATA)
    path = get_open_tracking_url_path(
        url, base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL)

    tracking_result = get_open_tracking_result(
        path, webhook_url=ALT_WEBHOOK_URL)
    assert tracking_result.tracked_url is None
    assert tracking_result.webhook_url == ALT_WEBHOOK_URL
    assert tracking_result.request_data is None
    assert tracking_result.metadata == ALT_METADATA
    assert tracking_result.is_open_tracking
    assert not tracking_result.is_click_tracking

def test_public_in_config_open_tracking_url_to_json():
    url = get_open_tracking_url(
        base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL,
        metadata=ALT_METADATA)
    path = get_open_tracking_url_path(
        url, base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL)

    tracking_result = get_open_tracking_result(
        path, webhook_url=ALT_WEBHOOK_URL).to_json_dict()
    assert tracking_result.tracked_url is None
    assert tracking_result.webhook_url == ALT_WEBHOOK_URL
    assert tracking_result.request_data is None
    assert tracking_result.metadata == ALT_METADATA
    assert tracking_result.is_open_tracking
    assert not tracking_result.is_click_tracking

def test_public_in_config_open_tracking_full_url():
    url = get_open_tracking_url(
        base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL,
        metadata=ALT_METADATA)

    tracking_result = get_open_tracking_result(
        url, webhook_url=ALT_WEBHOOK_URL,
        base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL)

    assert tracking_result.tracked_url is None
    assert tracking_result.webhook_url == ALT_WEBHOOK_URL
    assert tracking_result.request_data is None
    assert tracking_result.metadata == ALT_METADATA
    assert tracking_result.is_open_tracking
    assert not tracking_result.is_click_tracking

def test_public_embedded_open_tracking_url():
    url = get_open_tracking_url(
        base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL,
        webhook_url=ALT_WEBHOOK_URL,
        include_webhook_url=True,
        default_metadata=ALT_DEFAULT_METADATA,
        include_default_metadata=True,
        metadata=ALT_METADATA)
    path = get_open_tracking_url_path(
        url, base_open_tracking_url=ALT_BASE_OPEN_TRACKING_URL)

    tracking_result = get_open_tracking_result(
        path, request_data=ALT_REQUEST_DATA,
        include_default_metadata=True,
        include_webhook_url=True)

    assert tracking_result.tracked_url is None
    assert tracking_result.webhook_url == ALT_WEBHOOK_URL
    assert tracking_result.request_data == ALT_REQUEST_DATA
    assert tracking_result.metadata == ALT_EXPECTED_METADATA
    assert tracking_result.is_open_tracking
    assert not tracking_result.is_click_tracking

def test_public_basic_get_click_tracking_url():
    url = get_click_tracking_url(
        ALT_URL_TO_TRACK,
        base_click_tracking_url=ALT_BASE_CLICK_TRACKING_URL)
    assert url.startswith(ALT_BASE_CLICK_TRACKING_URL)
    assert "=" in url

def test_public_basic_get_click_tracking_url_append_slash():
    url = get_click_tracking_url(
        ALT_URL_TO_TRACK,
        base_click_tracking_url=ALT_BASE_CLICK_TRACKING_URL,
        append_slash=True)
    assert url.endswith("/")

def test_public_in_config_click_tracking_url():
    url = get_click_tracking_url(
        ALT_URL_TO_TRACK,
        base_click_tracking_url=ALT_BASE_CLICK_TRACKING_URL,
        metadata=ALT_METADATA)
    path = get_click_tracking_url_path(
        url, base_click_tracking_url=ALT_BASE_CLICK_TRACKING_URL)

    tracking_result = get_click_tracking_result(
        path, webhook_url=ALT_WEBHOOK_URL)
    assert tracking_result.tracked_url == ALT_URL_TO_TRACK
    assert tracking_result.webhook_url == ALT_WEBHOOK_URL
    assert tracking_result.request_data is None
    assert tracking_result.metadata == ALT_METADATA
    assert tracking_result.is_click_tracking
    assert not tracking_result.is_open_tracking

def test_public_in_config_click_tracking_full_url():
    url = get_click_tracking_url(
        ALT_URL_TO_TRACK,
        base_click_tracking_url=ALT_BASE_CLICK_TRACKING_URL,
        metadata=ALT_METADATA)

    tracking_result = get_click_tracking_result(
        url,
        webhook_url=ALT_WEBHOOK_URL,
        base_click_tracking_url=ALT_BASE_CLICK_TRACKING_URL)

    assert tracking_result.tracked_url == ALT_URL_TO_TRACK
    assert tracking_result.webhook_url == ALT_WEBHOOK_URL
    assert tracking_result.request_data is None
    assert tracking_result.metadata == ALT_METADATA
    assert tracking_result.is_click_tracking
    assert not tracking_result.is_open_tracking

File: RepoTransBench/target_projects/Python/C++/powergo_pytracking/public_tests/test_pytracking_public.cpp
#include <gtest/gtest.h>
#include "pytracking/configuration.h"
#include "pytracking/tracking.h"
using namespace pytracking;

const std::string ALT_URL_TO_TRACK = "https://anotherdomain.io/tracking/?data=newdata";
const std::string ALT_BASE_CLICK_TRACKING_URL = "https://y.z.com/track/";
const std::string ALT_BASE_OPEN_TRACKING_URL = "https://y.z.com/track/open/";
const std::string ALT_WEBHOOK_URL = "https://notify.me/webhook/";

std::map<std::string, util::Any> ALT_METADATA = {
    {"param4", "val4"},
    {"another", true},
    {"nested_alt", std::map<std::string, util::Any>{{"paramA", "valA"}}}
};
std::map<std::string, util::Any> ALT_DEFAULT_METADATA = {
    {"key42", false},
    {"strangeé", "winoèèè"},
    {"paramX", "other3"}
};
std::map<std::string, util::Any> ALT_EXPECTED_METADATA = []{
    auto m = ALT_DEFAULT_METADATA;
    m.insert(ALT_METADATA.begin(), ALT_METADATA.end());
    return m;
}();
std::map<std::string, util::Any> ALT_REQUEST_DATA = {
    {"user_agent", "Safari"},
    {"user_ip", "192.168.1.1"}
};

Configuration ALT_CONFIGURATION(
    ALT_WEBHOOK_URL,
    ALT_BASE_OPEN_TRACKING_URL,
    ALT_BASE_CLICK_TRACKING_URL,
    ALT_DEFAULT_METADATA
);

TEST(PublicPyTracking, PublicGetOpenTrackingPixel) {
    auto [pixel, mime] = get_open_tracking_pixel();
    EXPECT_TRUE(typeid(pixel) == typeid(std::vector<uint8_t>));
    EXPECT_EQ(mime, "image/png");
}
TEST(PublicPyTracking, PublicBasicGetOpenTrackingUrl) {
    auto url = get_open_tracking_url(ALT_BASE_OPEN_TRACKING_URL);
    EXPECT_TRUE(url.find(ALT_BASE_OPEN_TRACKING_URL) == 0);
    EXPECT_NE(url.substr(ALT_BASE_OPEN_TRACKING_URL.length()), "");
}
TEST(PublicPyTracking, PublicBasicGetOpenTrackingUrlAppendSlash) {
    auto url = get_open_tracking_url(ALT_BASE_OPEN_TRACKING_URL, true);
    EXPECT_TRUE(url.rfind("/") == (url.length() - 1));
}
TEST(PublicPyTracking, PublicInConfigOpenTrackingUrl) {
    auto url = get_open_tracking_url(ALT_BASE_OPEN_TRACKING_URL, false, ALT_METADATA);
    auto path = get_open_tracking_url_path(url, ALT_BASE_OPEN_TRACKING_URL);

    auto tracking_result = get_open_tracking_result(path, ALT_WEBHOOK_URL);
    EXPECT_EQ(tracking_result.tracked_url, std::nullopt);
    EXPECT_EQ(tracking_result.webhook_url, ALT_WEBHOOK_URL);
    EXPECT_EQ(tracking_result.request_data, std::nullopt);
    EXPECT_EQ(tracking_result.metadata, ALT_METADATA);
    EXPECT_TRUE(tracking_result.is_open_tracking);
    EXPECT_FALSE(tracking_result.is_click_tracking);
}
TEST(PublicPyTracking, PublicInConfigOpenTrackingUrlToJson) {
    auto url = get_open_tracking_url(ALT_BASE_OPEN_TRACKING_URL, false, ALT_METADATA);
    auto path = get_open_tracking_url_path(url, ALT_BASE_OPEN_TRACKING_URL);

    auto tracking_result = get_open_tracking_result(path, ALT_WEBHOOK_URL);
    auto result_json = tracking_result.to_json_dict();
    EXPECT_EQ(result_json["tracked_url"], nullptr);
    EXPECT_EQ(result_json["webhook_url"], ALT_WEBHOOK_URL);
    EXPECT_EQ(result_json["request_data"], nullptr);
    EXPECT_EQ(result_json["metadata"], ALT_METADATA);
    EXPECT_TRUE(result_json["is_open_tracking"]);
    EXPECT_FALSE(result_json["is_click_tracking"]);
}
TEST(PublicPyTracking, PublicInConfigOpenTrackingFullUrl) {
    auto url = get_open_tracking_url(ALT_BASE_OPEN_TRACKING_URL, false, ALT_METADATA);
    auto tracking_result = get_open_tracking_result(url, ALT_WEBHOOK_URL, ALT_BASE_OPEN_TRACKING_URL);
    EXPECT_EQ(tracking_result.tracked_url, std::nullopt);
    EXPECT_EQ(tracking_result.webhook_url, ALT_WEBHOOK_URL);
    EXPECT_EQ(tracking_result.request_data, std::nullopt);
    EXPECT_EQ(tracking_result.metadata, ALT_METADATA);
    EXPECT_TRUE(tracking_result.is_open_tracking);
    EXPECT_FALSE(tracking_result.is_click_tracking);
}
TEST(PublicPyTracking, PublicEmbeddedOpenTrackingUrl) {
    auto url = get_open_tracking_url(
        ALT_BASE_OPEN_TRACKING_URL, false, ALT_METADATA, ALT_WEBHOOK_URL, true,
        ALT_DEFAULT_METADATA, true
    );
    auto path = get_open_tracking_url_path(url, ALT_BASE_OPEN_TRACKING_URL);

    auto tracking_result = get_open_tracking_result(
        path, ALT_WEBHOOK_URL, std::nullopt, true, true, ALT_REQUEST_DATA
    );

    EXPECT_EQ(tracking_result.tracked_url, std::nullopt);
    EXPECT_EQ(tracking_result.webhook_url, ALT_WEBHOOK_URL);
    EXPECT_EQ(tracking_result.request_data, ALT_REQUEST_DATA);
    EXPECT_EQ(tracking_result.metadata, ALT_EXPECTED_METADATA);
    EXPECT_TRUE(tracking_result.is_open_tracking);
    EXPECT_FALSE(tracking_result.is_click_tracking);
}
TEST(PublicPyTracking, PublicBasicGetClickTrackingUrl) {
    auto url = get_click_tracking_url(ALT_URL_TO_TRACK, ALT_BASE_CLICK_TRACKING_URL);
    EXPECT_TRUE(url.find(ALT_BASE_CLICK_TRACKING_URL) == 0);
    EXPECT_NE(url.find("="), std::string::npos);
}
TEST(PublicPyTracking, PublicBasicGetClickTrackingUrlAppendSlash) {
    auto url = get_click_tracking_url(ALT_URL_TO_TRACK, ALT_BASE_CLICK_TRACKING_URL, true);
    EXPECT_TRUE(url.rfind("/") == (url.length() - 1));
}
TEST(PublicPyTracking, PublicInConfigClickTrackingUrl) {
    auto url = get_click_tracking_url(ALT_URL_TO_TRACK, ALT_BASE_CLICK_TRACKING_URL, false, ALT_METADATA);
    auto path = get_click_tracking_url_path(url, ALT_BASE_CLICK_TRACKING_URL);

    auto tracking_result = get_click_tracking_result(path, ALT_WEBHOOK_URL);
    EXPECT_TRUE(tracking_result.tracked_url.has_value() && tracking_result.tracked_url.value() == ALT_URL_TO_TRACK);
    EXPECT_EQ(tracking_result.webhook_url, ALT_WEBHOOK_URL);
    EXPECT_EQ(tracking_result.request_data, std::nullopt);
    EXPECT_EQ(tracking_result.metadata, ALT_METADATA);
    EXPECT_TRUE(tracking_result.is_click_tracking);
    EXPECT_FALSE(tracking_result.is_open_tracking);
}
TEST(PublicPyTracking, PublicInConfigClickTrackingFullUrl) {
    auto url = get_click_tracking_url(ALT_URL_TO_TRACK, ALT_BASE_CLICK_TRACKING_URL, false, ALT_METADATA);
    auto tracking_result = get_click_tracking_result(url, ALT_WEBHOOK_URL, ALT_BASE_CLICK_TRACKING_URL);

    EXPECT_TRUE(tracking_result.tracked_url.has_value() && tracking_result.tracked_url.value() == ALT_URL_TO_TRACK);
    EXPECT_EQ(tracking_result.webhook_url, ALT_WEBHOOK_URL);
    EXPECT_EQ(tracking_result.request_data, std::nullopt);
    EXPECT_EQ(tracking_result.metadata, ALT_METADATA);
    EXPECT_TRUE(tracking_result.is_click_tracking);
    EXPECT_FALSE(tracking_result.is_open_tracking);
}